api/envoy/config/listener/v3/listener.proto

```diff
@@ -90,6 +90,14 @@ message Listener {
   // Configuration for envoy internal listener. All the future internal listener features should be added here.
   // [#not-implemented-hide:]
   message InternalListenerConfig {
+    // The max contiguous chained internal listener to internal cluster is allowed. 0 means upstream
+    // connection cannot use internal connection. The goal is to avoid a bad config that containing
+    // loop of the internal listener and internal cluster. Such a loop will OOM the Envoy in a second.
+    // Per listener connection limit doesn't work well here because it takes 1 dispatcher cycle to reach
```

Comment:
How does this work when you connection Listener A -> Cluster C -> Listener B -> Cluster D and Listener B also specifies a `hop_to_live`? Do we only ever consider the root listener?

Is the default 1?

Reply:
We have two strategies to determine:
Lowest decide or origin decides.
if listener A allows 10 and listener B allow 5, when htl proporagate from A to B (now htl = 9), 
[1] the htl drop from 9 to 5 or
[2] the htl is untouched as 9 since origin had decided a value.

I don't have a preference. For safety, we would probably choose the lowest-decide. 
I would leave the default interpreted by PROTOBUF_GET_WRAPPED_OR_DEFAULT, again for safety,
the default value is 1.

Does the above make sense? If so I will update the description.

Reply:
Are there uses for hop_to_live > 1?  The cost per client stream grows significantly every time there is recursion through the system.  It would be good to have a clear use case for higher number of hops before we decide to allow Envoy to operate in that mode.

Reply:
Hypothetical example would be a CONNECT tunnel wrapping another CONNECT tunnel. I think it would help for something less hypothetical though.

Reply:
I didn't find the case so I feel it's good to start with default value == 1 (field is empty in protobuf message) and leave recursive possible.

Reply:
CONNECT over CONNECT seems like something we don't want to support.

Reply:
@antoniovicente Can you clarify the `operate mode`? I am uncertain what you are requesting
1. Default value? I think we all agree 1 as default value here
2. Cost of accounting? 
3. Reject config?

Reply:
I suggest we not have a parameter until we have a clear case for "hop_to_live > 1"

Just enforce that there's at most a single internal listener when accepting a request.

Reply:
I see. Make perfect sense!

How about the underlying accounting by a per stream hop_to_live?

Reply:
Being able to detect during config load would help avoid black holing of traffic.  I acknowledge that there are some challenges, but ideally we can verify during CDS load that a cluster that points to an internal listener is not configured behind an internal listener.

Doing the check in the data plane is fine, it is just somewhat riskier.  Configs that reject all requests will not be detected until traffic arrives and all requests are rejected.

Reply:
> Being able to detect during config load would help avoid black holing of traffic. I acknowledge that there are some challenges, but ideally we can verify during CDS load that a cluster that points to an internal listener is not configured behind an internal listener.

I don't see how this is possible given eventual consistency. If the referenced listener is not yet loaded by LDS we can't know if it's an internal listener yet. The only way to achieve this in xDS would be to have a distinct xDS type for internal listeners, which would require totally reforming the listener resource type model (possibly, but major work unrelated to this change). The best we can do today is to do this on data plane AFAICT.

CC @envoyproxy/api-shepherds

Reply:
I agree with @htuch that we probably must do data plane detection, though I think this is OK given that we do similar things already like CDN loop detection. I also agree with @antoniovicente that I would rather just hard code this to 1 in the code and avoid config until there is an actual use case.